### ISSUE TITLE
Windows Server 2016 on Azure cannot start docker service via build script

### ISSUE THREAD

**Issue by heynickc** [MicrosoftDocs/Virtualization-Documentation#520]
Hello,

I've been dealing with an issue in a scenario where we are using Docker inside of a TeamCity build agent to run integration tests.  TeamCity deploys VMs from SysPrep'd VHDs, for which I've set up a base image that has Windows Server 2016 Containers, Docker running, base images installed, etc. everything working fine when RDP'd into the VM.

The issue happens when a new VM is deployed for the TeamCity build.  Firstly, the docker service doesn't start automatically.  Since the VM is deployed dynamically to handle certain CI loads, I cannot RDP into the server and run `Start-Service docker`.  When I create an image manually, RDP into it, all I simply have to do is run `Start-Service docker`, the docker daemon starts, and everything is fine.  However, since the VMs are deployed dynamically, I cannot do this via an interactive Powershell session.

So what I'm needing to do is script the starting of the docker service via FAKE, using a Powershell command that is created via System.Management.Automation namespace.  However, running `Start-Service docker` in a non-interactive Powershell session does not work.  I've tried both the `Debug-ContainerHost.ps1` and also looked at the event logs and see the following entry:
```
"Error","fatal: Error starting daemon: Error initializing network controller: hnsCall failed in Win32: Server execution failed (0x80080005)"
```
Here is the output from my FAKE build script via TeamCity, including the output of the Debug-ContainerHost.ps1 and the event logs for docker.  I understand that I don't meet the requirement for "Has KB3192366, KB3194496, or later installed..." however being that I can resolve the issue via interactive Powershell from an RDP, I'm not sure if that is the root cause:
```
[19:25:16][Step 1/1] RestartDocker (2m:23s)
[19:25:16][RestartDocker] Target: RestartDocker
[19:25:16][RestartDocker] Starting Target: RestartDocker (==> ActivateFinalTargets)
[19:25:16][RestartDocker] powershell.exe Invoke-WebRequest https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/master/windows-server-container-tools/Debug-ContainerHost/Debug-ContainerHost.ps1 -UseBasicParsing | Invoke-Expression
[19:25:39][RestartDocker] Checking for common problems
[19:25:39][RestartDocker] Describing Windows Version and Prerequisites
[19:25:39][RestartDocker]  [+] Is Windows 10 Anniversary Update or Windows Server 2016 475ms
[19:25:39][RestartDocker]  [-] Has KB3192366, KB3194496, or later installed if running Windows build 14393 246ms
[19:25:39][RestartDocker]    Expected {321} to be greater than or equal to {351}
[19:25:39][RestartDocker]    13:             (Get-ItemProperty -Path 'HKLM:\software\Microsoft\Windows NT\CurrentVersion' -Name UBR).UBR | Should Not BeLessThan 351
[19:25:39][RestartDocker]    at <ScriptBlock>, <No file>: line 13
[19:25:39][RestartDocker]  [+] Is not a build with blocking issues 108ms
[19:25:39][RestartDocker] Describing Docker is installed
[19:25:39][RestartDocker]  [+] A Docker service is installed - 'Docker' or 'com.Docker.Service'  109ms
[19:25:39][RestartDocker]  [-] Service is running 31ms
[19:25:39][RestartDocker]    Expected: {True}
[19:25:39][RestartDocker]    But was:  {False}
[19:25:39][RestartDocker]    39:         $AtLeastOneRunning | Should Be $true
[19:25:39][RestartDocker]    at <ScriptBlock>, <No file>: line 39
[19:25:39][RestartDocker]  [+] Docker.exe is in path 3.12s
[19:25:39][RestartDocker]  [+] Docker is registered in the EventLog service 24ms
[19:25:39][RestartDocker] Describing User has permissions to use Docker daemon
[19:25:39][RestartDocker]  [+] docker.exe should not return access denied 29ms
[19:25:39][RestartDocker] Describing Windows container settings are correct
[19:25:39][RestartDocker]  [+] Do not have DisableVSmbOplock set to 1 51ms
[19:25:39][RestartDocker]  [+] Do not have zz values set 268ms
[19:25:39][RestartDocker] Describing The right container base images are installed
[19:25:39][RestartDocker]  [-] At least one of 'microsoft/windowsservercore' or 'microsoft/nanoserver' should be installed 590ms
[19:25:39][RestartDocker]    ValidationMetadataException: The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
[19:25:39][RestartDocker]    ParameterBindingValidationException: Cannot validate argument on parameter 'Property'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
[19:25:39][RestartDocker]    at <ScriptBlock>, <No file>: line 99
[19:25:39][RestartDocker] Describing Container network is created
[19:25:39][RestartDocker]  [-] Error occurred in Describe block 1.17s
[19:25:39][RestartDocker]    RuntimeException: Cannot index into a null array.
[19:25:39][RestartDocker]    at <ScriptBlock>, <No file>: line 128
[19:25:39][RestartDocker] Showing output from: docker info
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] Showing output from: docker version
[19:25:39][RestartDocker] Client:
[19:25:39][RestartDocker]  Version:      1.12.2-cs2-ws-beta
[19:25:39][RestartDocker]  API version:  1.25
[19:25:39][RestartDocker]  Go version:   go1.7.1
[19:25:39][RestartDocker]  Git commit:   050b611
[19:25:39][RestartDocker]  Built:        Tue Oct 11 02:35:40 2016
[19:25:39][RestartDocker]  OS/Arch:      windows/amd64
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] Showing output from: docker network ls
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] Warnings & errors from the last 24 hours
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker]    ProviderName: docker
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] TimeCreated                     Id LevelDisplayName Message                    
[19:25:39][RestartDocker] -----------                     -- ---------------- -------                    
[19:25:39][RestartDocker] 12/19/2016 7:18:46 PM            4 Error            fatal: Error starting da...
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] 
[19:25:39][RestartDocker] Logs saved to D:\work\46a1dbd8d00c38f\logs_20161219-192538.csv
[19:25:39][RestartDocker] Get-ItemProperty : Cannot find path 'HKLM:\SOFTWARE\Microsoft\Windows 
[19:25:39][RestartDocker] NT\CurrentVersion\Virtualization\Containers' because it does not exist.
[19:25:39][RestartDocker] At line:79 char:23
[19:25:39][RestartDocker] + ... regvalues = Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows  ...
[19:25:39][RestartDocker] +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[19:25:39][RestartDocker]     + CategoryInfo          : ObjectNotFound: (HKLM:\SOFTWARE\...tion\Containe 
[19:25:39][RestartDocker]    rs:String) [Get-ItemProperty], ItemNotFoundException
[19:25:39][RestartDocker]     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetIt 
[19:25:39][RestartDocker]    emPropertyCommand
[19:25:39][RestartDocker]  
[19:25:39][RestartDocker] Get-Member : You must specify an object for the Get-Member cmdlet.
[19:25:39][RestartDocker] At line:80 char:24
[19:25:39][RestartDocker] +          ($regvalues | Get-Member zz* | Measure-Object).Count | Shoul ...
[19:25:39][RestartDocker] +                        ~~~~~~~~~~~~~~
[19:25:39][RestartDocker]     + CategoryInfo          : CloseError: (:) [Get-Member], InvalidOperationEx 
[19:25:39][RestartDocker]    ception
[19:25:39][RestartDocker]     + FullyQualifiedErrorId : NoObjectInGetMember,Microsoft.PowerShell.Command 
[19:25:39][RestartDocker]    s.GetMemberCommand
[19:25:39][RestartDocker]  
[19:25:39][RestartDocker] error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/images/json: open //./pipe/docker_engine: The system cannot find the file specified.
[19:25:39][RestartDocker] True
[19:25:39][RestartDocker] #TYPE System.Diagnostics.EventLogEntry#Application/docker/1
[19:25:39][RestartDocker] "EventID","MachineName","Data","Index","Category","CategoryNumber","EntryType","Message","Source","ReplacementStrings","InstanceId","TimeGenerated","TimeWritten","UserName","Site","Container"
[19:25:39][RestartDocker] "1","tcimagepb","System.Byte[]","2171","(0)","0","Information","Graph migration to content-addressability took 0.00 seconds","docker","System.String[]","1","12/19/2016 7:14:07 PM","12/19/2016 7:14:07 PM",,,
[19:25:39][RestartDocker] "1","tcimagepb","System.Byte[]","2172","(0)","0","Information","Loading containers: start.","docker","System.String[]","1","12/19/2016 7:14:07 PM","12/19/2016 7:14:07 PM",,,
[19:25:39][RestartDocker] "1","tcimagepb","System.Byte[]","2165","(0)","0","Information","Windows default isolation mode: process","docker","System.String[]","1","12/19/2016 7:14:05 PM","12/19/2016 7:14:05 PM",,,
[19:25:39][RestartDocker] "1","tcimagepb","System.Byte[]","2166","(0)","0","Information","[graphdriver] using prior storage driver: windowsfilter","docker","System.String[]","1","12/19/2016 7:14:05 PM","12/19/2016 7:14:05 PM",,,
[19:25:39][RestartDocker] "1","PBW1541","System.Byte[]","2178","(0)","0","Information","Windows default isolation mode: process","docker","System.String[]","1","12/19/2016 7:16:45 PM","12/19/2016 7:16:45 PM",,,
[19:25:39][RestartDocker] "1","PBW1541","System.Byte[]","2187","(0)","0","Information","Loading containers: start.","docker","System.String[]","1","12/19/2016 7:16:46 PM","12/19/2016 7:16:46 PM",,,
[19:25:39][RestartDocker] "4","PBW1541","System.Byte[]","2191","(0)","0","Error","fatal: Error starting daemon: Error initializing network controller: hnsCall failed in Win32: Server execution failed (0x80080005)","docker","System.String[]","4","12/19/2016 7:18:46 PM","12/19/2016 7:18:46 PM",,,
[19:25:39][RestartDocker] "1","PBW1541","System.Byte[]","2179","(0)","0","Information","[graphdriver] using prior storage driver: windowsfilter","docker","System.String[]","1","12/19/2016 7:16:45 PM","12/19/2016 7:16:45 PM",,,
[19:25:39][RestartDocker] "1","PBW1541","System.Byte[]","2186","(0)","0","Information","Graph migration to content-addressability took 0.00 seconds","docker","System.String[]","1","12/19/2016 7:16:46 PM","12/19/2016 7:16:46 PM",,,
[19:25:39][RestartDocker] Starting Service docker on PBW1541
[19:25:39][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:40][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:41][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:42][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:43][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:44][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:45][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:46][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
[19:25:47][RestartDocker] Waiting for docker to start (Timeout: 00:02:00)
... continue waiting
[19:27:39][RestartDocker] Running build failed.
[19:27:39][RestartDocker] Error:
[19:27:39][RestartDocker] System.Exception: The service docker has not been started (check the logs for errors)
[19:27:39][RestartDocker]    at Fake.ServiceControllerHelpers.ensureRemoteServiceHasStarted@115-3.Invoke(String message) in C:\code\FAKE\src\app\FakeLib\ServiceControllerHelper.fs:line 115
[19:27:39][RestartDocker]    at Fake.ServiceControllerHelpers.ensureRemoteServiceHasStarted(String host, String name, TimeSpan timeout) in C:\code\FAKE\src\app\FakeLib\ServiceControllerHelper.fs:line 115
[19:27:39][RestartDocker]    at FSI_0005.Build.clo@145-14.Invoke(Unit _arg8) in D:\work\46a1dbd8d00c38f\build.fsx:line 175
[19:27:39][RestartDocker]    at Fake.TargetHelper.runSingleTarget(TargetTemplate`1 target) in C:\code\FAKE\src\app\FakeLib\TargetHelper.fs:line 492
[19:27:39][RestartDocker] Starting FinalTarget: TearDownDbContainer
```
The error message about the hnsCall seems to be the common thread here and there is another github issue filed under docker/docker here:  https://github.com/docker/docker/issues/27708.

I have tried the suggested fix (remove registry key `HKLM:\SYSTEM\CurrentControlSet\Control\Wininit\Headless`) from that github issue and it hasn't resolved the issue.  I wanted to raise it here to get some more eyes on it, in case someone had any more suggestions or workarounds.

Thanks for your help.

**Comment by PatrickLang** [comment]
If you update to a recent KB (November or later), then the Headless fix is already included. You don't need to change any regkeys.

Have you tried this with a new KB yet? If so, please post the updated Debug-ContainerHost.ps1 output so we can check it out.

There's also a chance that you're missing the containers role since that regkey is missing. See #487 for a bug in Debug-ContainerHost.ps1

Thanks!

**Comment by gaurarpit12** [comment]
I am trying to run a docker command from VSTS 2017 on Windows Server 2016 Azure machine, under the "Run a Docker Command" build step.
But I am getting the following error :-

error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.25/containers/json?all=1: open //./pipe/docker_engine: Access is denied.

Please help me out with that.

**Comment by PatrickLang** [comment]
ping @glennc do you have any resources on building from VSTS?

**Comment by PatrickLang** [comment]
@glennc has done demos using VSTS to build in containers, and also build container images

Here's his session from the MS Ignite event:
https://myignite.microsoft.com/videos/3122

**Comment by gaurarpit12** [comment]
@PatrickLang he has used linux environment there. But I am using Windows environment. 
I don't have much idea of Linux, so help me with that.

**Comment by gaurarpit12** [comment]
@PatrickLang Could you please reply with some solution?
@glennc could you help ?

**Comment by friism** [comment]
@gaurarpit12 can you invoke normal `docker` commands like `docker images`? Depending on how the Docker engine was installed, if the user running Visual Studio is not in the administrator group, you would get "access denied": https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon#set-docker-security-group

**Comment by PatrickLang** [comment]
Looking at the logs above, I think it's going to be best to:
- Update to the latest Windows update. Lots of bugs have been fixed since 14393.321
- Make sure your IP on the host doesn't start with 172.*
- Manually install Docker v1.13

That will get all the latest available fixes to Windows & Docker. Once that's done, try starting the Docker service then running the Debug-ContainerHost.ps1 script again to see what it finds.

**Comment by mylesbkeating1993** [comment]
@gaurarpit12 did @PatrickLang's suggestions help out?  Anything else we can help with?

**Comment by enderb-ms** [comment]
Closing this issue due to inactivity. Please feel free to open this again if there is still something we can help with.

**Comment by Cloudmersive** [comment]
This issue needs to be reopened because it was never fixed

**Comment by Cloudmersive** [comment]
> Looking at the logs above, I think it's going to be best to:
> 
> * Update to the latest Windows update. Lots of bugs have been fixed since 14393.321 including HNS
> * Make sure your IP on the host doesn't start with 172.*
> * Manually install Docker v1.13
> 
> That will get all the latest available fixes to Windows & Docker. Once that's done, try starting the Docker service then running the Debug-ContainerHost.ps1 script again to see what it finds.

This suggestion does not make sense because the latest version of Docker is version 18+ not 1.x
